### ISSUE TITLE
Matrix size incompatible error when using nested gradient tape

### ISSUE THREAD

**Issue by maximocrv** [apple/tensorflow_macos#167]
I get an error that did not previously occur on my intel mac using TF 2.2.0 with the following loss function:

```
def get_grad_and_loss(self, x, y):
        with tf.GradientTape(persistent=True) as gl_tape:
            gl_tape.watch(x)

            with tf.GradientTape(persistent=True) as l_tape:
                l_tape.watch(x)
                y_pred = self.call(x)

            grad_mat = l_tape.gradient(y_pred, x)
            loss = tf.reduce_mean(tf.math.square(y_pred - y[:, tf.newaxis])) + tf.reduce_mean(tf.maximum(0, -1 * (grad_mat[:, 0])))

        g = gl_tape.gradient(loss, self.trainable_weights)

        return g, loss
```

The error message is `OP_REQUIRES failed at mlc_matmul_ops.cc:70 : Invalid argument:GetOutputShape: Matrix size-incompatible: In[0]: [256,4], In[1]: [4,400]`. Any idea how to address this issue?

**Comment by anna-tikhonova** [comment]
Thank you very much for reporting this issue. Could you please provide a reproducible test case, so we can run it locally and investigate?

**Comment by maximocrv** [comment]
Sure here you go.

```
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras import initializers
from tensorflow.keras.optimizers import Adam

class MyModel(tf.keras.Model):
    def __init__(self, input_dim, mode, activation='relu', sc_weights=(1, 1), epsilon=0):
        super(MyModel, self).__init__()
        self.input_dim = input_dim
        self.mode = mode
        self.sc_weights = sc_weights
        self.epsilon = epsilon
        self.dense1 = Dense(400, activation=activation, kernel_initializer=initializers.glorot_uniform(),
                            input_dim=self.input_dim)
        self.dense2 = Dense(400, activation=activation, kernel_initializer=initializers.glorot_uniform())
        self.dense3 = Dense(400, activation=activation, kernel_initializer=initializers.glorot_uniform())
        self.dense4 = Dense(400, activation=activation, kernel_initializer=initializers.glorot_uniform())
        self.dense_out = Dense(1, activation=activation, kernel_initializer=initializers.glorot_uniform())

    def call(self, inputs, **kwargs):
        x = self.dense1(inputs)
        x = self.dense2(x)
        x = self.dense3(x)
        x = self.dense4(x)
        return self.dense_out(x)

    def get_grad_and_loss(self, x, y):
        with tf.GradientTape(persistent=True) as gl_tape:
            gl_tape.watch(x)

            with tf.GradientTape(persistent=True) as l_tape:
                l_tape.watch(x)
                y_pred = self.call(x)

            grad_mat = l_tape.gradient(y_pred, x)
            loss = tf.reduce_mean(tf.math.square(y_pred - y[:, tf.newaxis])) \
                   + self.sc_weights[0] * tf.reduce_mean(tf.maximum(0, -1 * (grad_mat[:, 0]))) \
                   + self.sc_weights[1] * tf.reduce_mean(tf.maximum(0, 1 * grad_mat[:, 1]))

        g = gl_tape.gradient(loss, self.trainable_weights)

        return g, loss

x_train = np.random.randn(500, 4).astype(np.float32)
y_train = np.random.randn(500).astype(np.float32)
train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
train_dataset = train_dataset.shuffle(buffer_size=1024).batch(64)

model = MyModel(input_dim=4, mode='_')

epochs = 10
for epoch in range(epochs):
    print(f'############ START OF EPOCH {epoch + 1} ################')
    for step, (x_batch_train, y_batch_train) in enumerate(train_dataset):
        grads, loss = model.get_grad_and_loss(x_batch_train, y_batch_train)
        Adam.apply_gradients(zip(grads, model.trainable_weights))
```

**Comment by maximocrv** [comment]
As a sidenote I am able to run this without any errors using the non macos version of TF 2.4 on collab so I think there is some issue here with the MLCmatmulgrad functionality.
